70_FLIPFluids/main.cpp

```diff
@@ -0,0 +1,2240 @@
+		
+		WorkgroupCountParticles = (numParticles + WorkgroupSize - 1) / WorkgroupSize;
+		WorkgroupCountGrid = (numGridCells + WorkgroupSize - 1) / WorkgroupSize;
+
+		{
+			float zNear = 0.1f, zFar = 10000.f;
+			core::vectorSIMDf cameraPosition(14, 8, 12);
+			core::vectorSIMDf cameraTarget(0, 0, 0);
+			matrix4SIMD projectionMatrix = matrix4SIMD::buildProjectionMatrixPerspectiveFovLH(core::radians(60.0f), float(WIN_WIDTH) / WIN_HEIGHT, zNear, zFar);
+			camera = Camera(cameraPosition, cameraTarget, projectionMatrix, 1.069f, 0.4f);
+
+			m_pRenderParams.zNear = zNear;
+			m_pRenderParams.zFar = zFar;
+		}
+		m_pRenderParams.radius = m_gridData.gridCellSize * 0.4f;
+
+		// create buffers
+		video::IGPUBuffer::SCreationParams params = {};
+		params.size = sizeof(SGridData);
+		params.usage = IGPUBuffer::EUF_UNIFORM_BUFFER_BIT | IGPUBuffer::EUF_TRANSFER_DST_BIT | IGPUBuffer::EUF_INLINE_UPDATE_VIA_CMDBUF;
+		createBuffer(gridDataBuffer, params);
+
+		params.size = 2 * sizeof(float32_t4);
+		createBuffer(pressureParamsBuffer, params);
+		
+		params.size = sizeof(SMVPParams);
+		createBuffer(cameraBuffer, params);
+
+		params.size = sizeof(SParticleRenderParams);
+		createBuffer(pParamsBuffer, params);
+
+		params.size = numParticles * sizeof(Particle);
+		params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT | IGPUBuffer::EUF_TRANSFER_SRC_BIT;
+		createBuffer(particleBuffer, params);
+		params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT | IGPUBuffer::EUF_TRANSFER_DST_BIT;
+		createBuffer(tempParticleBuffer, params);
+
+		params.size = numGridCells * sizeof(uint32_t2);
+		params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT | IGPUBuffer::EUF_TRANSFER_DST_BIT;
+		createBuffer(gridParticleIDBuffer, params);
+
+		params.size = numGridCells * sizeof(float);
+		params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT | IGPUBuffer::EUF_TRANSFER_DST_BIT;
+		createBuffer(pressureBuffer, params);
+		params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT | IGPUBuffer::EUF_TRANSFER_SRC_BIT;
+		createBuffer(tempPressureBuffer, params);
+
+		params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT;
+		createBuffer(divergenceBuffer, params);
+
+		params.size = numGridCells * sizeof(uint32_t);
+		params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT;
+		createBuffer(gridCellMaterialBuffer, params);
+		createBuffer(tempCellMaterialBuffer, params);
+
+		params.size = numGridCells * sizeof(uint32_t4);
+		createBuffer(gridAxisCellMaterialBuffer, params);
+		createBuffer(tempAxisCellMaterialBuffer, params);
+
+		params.size = numGridCells * sizeof(float32_t4);
+		createBuffer(gridDiffusionBuffer, params);
+		createBuffer(tempDiffusionBuffer, params);
+
+		params.size = numParticles * 6 * sizeof(VertexInfo);
+		createBuffer(particleVertexBuffer, params);
+
+		params.size = numParticles * sizeof(uint32_t2);
+		createBuffer(particleCellPairBuffer, params);
+
+		// velocity field stuffs
+		{
+			IGPUImage::SCreationParams imgInfo;
+			imgInfo.format = asset::EF_R32G32B32A32_SFLOAT;
```

Comment:
experiment with half-float format

Reply:
btw for the atomics, one of the image arrays (the one you'll accumulate particle velocities into) needs to stay as `R32_SFLOAT` because there's no atomic capability on 16bit format images (float or uint16)